kernel/src/engine/default/storage.rs

```diff
@@ -1,38 +1,141 @@
-#[cfg(feature = "cloud")]
-use hdfs_native_object_store::HdfsObjectStore;
 use object_store::parse_url_opts as parse_url_opts_object_store;
 use object_store::path::Path;
 use object_store::{Error, ObjectStore};
 use url::Url;
 
+use std::collections::HashMap;
+use std::sync::{Arc, OnceLock, RwLock};
+
+/// Alias for convenience
+type ClosureReturn = Result<(Box<dyn ObjectStore>, Path), Error>;
+/// This type alias makes it easier to reference the handler closure(s)
+///
+/// It uses a HashMap<String, String> which _must_ be converted in our [parse_url_opts] because we
+/// cannot use generics in this scenario.
+type HandlerClosure = Arc<dyn Fn(&Url, HashMap<String, String>) -> ClosureReturn + Send + Sync>;
+/// hashmap containing scheme => handler fn mappings to allow consumers of delta-kernel-rs provide
+/// their own url opts parsers for different scemes
+type Handlers = HashMap<String, HandlerClosure>;
+/// The URL_REGISTRY contains the custom URL scheme handlers that will parse URL options
+static URL_REGISTRY: OnceLock<RwLock<Handlers>> = OnceLock::new();
+
+/// Return to fully initialied URL handlers to allow extensible provisioning of [object_store] URL
+/// handlers
+fn get_url_handlers<'life>() -> &'life RwLock<Handlers> {
+    URL_REGISTRY.get_or_init(|| RwLock::new(HashMap::default()))
+}
```

Comment:
`LazyLock` might simplify this?
```rust
/// The URL_REGISTRY contains the custom URL scheme handlers that will parse URL options
static URL_REGISTRY: Lazy<RwLock<Handlers>> = Lazy::new(|| RwLock::new(HashMap::default()));
```

Also do we need RwLock if we can somehow enforce no modification after initialization?

Reply:
There has to be modification after initialization, the code must be initialized in kernel, and other callsites outside of kernel will be able to invoke it, we cannot control _when- that happens.